### ISSUE TITLE
HTML5 <details>

### ISSUE THREAD

**Issue by jrappen** [Python-Markdown/markdown#543]
Having an issue where:

```markdown
<details>
<summary>Some summary</summary>

...content either md or html if separated by new line from html tags

</details>
```

breaks a [Sublime Text](https://www.sublimetext.com) [`mdpopups`](https://github.com/facelessuser/sublime-markdown-popups) phantom from the first `<details>` tag on. Everything before that is fine.

---

* compare facelessuser/sublime-markdown-popups#25

**Comment by waylan** [comment]
We only recognize known block level tags as raw HTML blocks. All other tags are treated as inline HTML. I don't see where the HTML spec even mentions a `<details>` tag, let alone define it as block level.

**Comment by waylan** [comment]
Hmm, so I found [this][1], but the link it provides to the spec does not point to a `<details>` tag (it appears to be an old broken link) and the [Table of Contents][2] makes no mention of such a tag. As far as I can tell, its not in the spec, which is why its not supported by Python-Markdown. If you can point to where it is in the spec, then we'll add it, otherwise we won't.

[1]: https://www.w3.org/wiki/HTML/Elements/details
[2]: https://www.w3.org/TR/html5/Overview.html#contents

**Comment by jrappen** [comment]
Probably a block element cause you can nest them?

* compare <http://jsbin.com/egefop/14#html,live>

---

Also, you can pre-open them.
```markdown
<details open>
  <summary>Some details</summary>
  <p>More info about the details.</p>
</details>
```

* <http://www.w3schools.com/tags/tag_details.asp>
* <https://developer.mozilla.org/de/docs/Web/HTML/Element/details>
* <http://html5doctor.com/the-details-and-summary-elements/>

**Comment by jrappen** [comment]
Not in the specs as far as I can tell.

**Comment by facelessuser** [comment]
I found this https://dev.w3.org/html5/spec-preview/the-details-element.html.  It doesn't specifically say it is a block, but from context, it seems that is exactly what it is.

I have no idea if it is an official tag, or one that was proposed at one time and some browsers adopted it.

**Comment by jrappen** [comment]
Seems to have been added in `HTML5.1`?

* <https://www.w3.org/TR/html/interactive-elements.html#the-details-element>

**Comment by waylan** [comment]
I found it. It is in the new (draft) [HTML 5.2 spec][1]. As that spec is in draft form, I am unsure of whether we should add support now or wait. Anyone know if there is a summary of changes/new tags added in the 2.5 spec?

[1]: http://w3c.github.io/html/interactive-elements.html#the-details-element

**Comment by waylan** [comment]
Sigh. I haven't been following the spec since HTML 5.0 was finalized. Apparently there are multiple minor releases in the works with each one adding new tags. I really am not interested in constantly needing to add tags to the list of block level tags. We need a better solution. I'll have to think on this. A couple possible solutions come to mind:

1. Import some existing library which maintains a current list of block level tags. I'd prefer something in the Standard Library, but that is likely to move even slower that we would.
2. Reference some list published by others of only the new tags introduced in each minor version of the spec. First, it needs to exist and then we need to find it.

**Comment by facelessuser** [comment]
Yes. It would definitely be more ideal if there was a way to one shot this issue in a way to get all known block elements.

**Comment by facelessuser** [comment]
I get the feeling this is will be hard to define moving forward as elements are fundamentally defined differently under HTML5.  From what I gather from reading HTML5, elements aren't really defined by **block** content anymore, but are defined by **flow** content.   **Inline** seems to be denoted by **phrasing** content.  Furthermore, things like the anchor tag are actually [**transparent**](http://w3c.github.io/html-reference/spec.html#a-changes), which means they can hold **flow** content as well, in certain situations.  Because why not have ambiguous elements :angry: .

I guess this moves into what do the browsers default them to.  In the case of details, it is [block](http://w3c.github.io/html-reference/spec.html#details-display).

---

I don't think most parsers distinguish between block vs inline as they don't care, so finding a parser to do this for us might not be possible.

Mozilla seems to list [**block** level elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Block-level_elements), but you will notice that `details` is not listed under block.  This is because HTML5 no longer uses those terms.  But it *is* under [**flow** content](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Flow_content) (but so is the anchor tag :frowning:).

---

So in short  ¯\\\_(ツ)\_/¯.

**Comment by facelessuser** [comment]
Seems like I am still wrong as flow seems to contain everything.  Anyways, the point is that the model is more complex.

**Comment by waylan** [comment]
The relevant part of the Markdown [rules][1] state (emphasis added):

[1]: http://daringfireball.net/projects/markdown/syntax#html

> The only restrictions are that block-level HTML elements — e.g. `<div>`, `<table>`, `<pre>`, `<p>`, etc. — must be separated from surrounding content by blank lines, and the start and end tags of the block should not be indented with tabs or spaces. ***Markdown is smart enough not to add extra (unwanted) `<p>` tags around HTML block-level tags.***
>
> ...
>
> Note that Markdown formatting syntax is not processed within block-level HTML tags. E.g., you can't use Markdown-style `*emphasis*` inside an HTML block.
>
> Span-level HTML tags — e.g. `<span>`, `<cite>`, or `<del>` — can be used anywhere in a Markdown paragraph, list item, or header. If you want, you can even use HTML tags instead of Markdown formatting; e.g. if you'd prefer to use HTML `<a>` or `<img>` tags instead of Markdown's
link or image syntax, go right ahead. 
>
> Unlike block-level HTML tags, Markdown syntax *is* processed within span-level tags.

So the question is, what in the HTML5 spec maps to "block-level HTML tags" in the Markdown rules? When it has been unclear, I have usually used the rule that a `<p>` tag cannot contain any block-level tags, so I would classify anything which can be a child of a `<p>` tag to not be block-level. Note that the [HTML 5.2 spec][2] states:

> A `<p>` element’s end tag may be omitted if the `<p>` element is immediately followed by an `<address>`, `<article>`, `<aside>`, `<blockquote>`, `<details>`, `<div>`, `<dl>`, `<fieldset>`, `<figcaption>`, `<figure>`, `<footer>`, `<form>`, `<h1>`, `<h2>`,`<h3>`, `<h4>`, `<h5>`, `<h6>`, `<header>`, `<hr>`, `<main>`, `<menu>`, `<nav>`, `<ol>`, `<p>`,`< pre>`, `<section>`, `<table>`, or `<ul>`, element, or if there is no more content in the parent element and the parent element is an HTML element that is not an `<a>`, `<audio>`, `<del>`, `<ins>`, `<map>`, `<noscript>`, or `<video>` element.

I think that is a pretty reasonable list to map to "block-level HTML tags". And note that it explicitly eliminates the `<a>` tag (among others) from the list even though it is "flow content". Relevant to this issue, the `<details>` tag is included in that list.

For comparison, the [HTML5.0 spec ][3] gave this list:

> A `p` element's end tag may be omitted if the `p` element is immediately followed by an `address`, `article`, `aside`, `blockquote`, `div`, `dl`, `fieldset`, `footer`, `form`, `h1`, `h2`, `h3`, `h4`, `h5`, `h6`, `header`, `hgroup`, `hr`, `main`, `nav`, `ol`, `p`, `pre`, `section`, `table`, or `ul`, element, or if there is no more content in the parent element and the parent element is not an `a` element.

In addition to that list we also define `script`, `noscript`, `fieldset`, `iframe`, `math`, `style`, `thead`, `tbody`, `tr`, `th`, `td`, `group`, `canvas`, `output`, `progress`, and `video` tags as "block-level". I think those all make sense. Although, according to the HTML5.2 spec, it looks like we could arguably remove `noscript` and `video` tags from the list but I don't think we want Markdown to insert those into a `p` tag so we'll leave them as-is.

[2]: http://w3c.github.io/html/grouping-content.html#the-p-element
[3]: https://www.w3.org/TR/html5/grouping-content.html#the-p-element

**Comment by waylan** [comment]
In my previous comment I forgot to mention one additional point. Among other tags,  the `<math>` and `<canvas>` tags could be wrapped in a `<p>` elements, but we definitely don't want Markdown messing with their contents. If the user really wants those tags wrapped in a `<p>` tag, then they can do that themselves with raw HTML.

Therefore, for us to classify an HTML tag as "block-level", the element needs to meet at least one of the following conditions:

1. It is invalid to wrap the element in a `<p>` tag.
2. Markdown should not be mucking up the contents of the element

Note that those correspond with the two differences between how Markdown handles "block-level" versus "span-level" raw HTML elements.

**Comment by facelessuser** [comment]
Nice! Thanks for explaining what the approach is.  It's good to understand for the future.  I hadn't realized it was relative to `<p>`.  That gives us an easy way moving forward to tell what we can add to the block list.   Special considerations can be brought up separately.

**Comment by facelessuser** [comment]
FYI. I threw together a plugin called [Spoilers](http://facelessuser.github.io/pymdown-extensions/extensions/spoilers/) that inserts `<details><summary>`.  Basically it borrows syntax from admonitions and injects `<details><summary>` in either a close or open state by default with an optional CSS class name.  You generally need a details polyfill (as simple or complex as your use case requires) due to poor browser support currently.

Anyways.  You should get `<details><summary>` with nested markdown, and you can even style them with special classes.

This does not change things for `sublime-markdown-popups` as Sublime can't handle `<details>` anyway, and we have no way of injecting JavaScript into `minihtml` to change that.  So I don't plan on even including Spoilers in `sublime-markdown-popups`  But for general convenience with working with `<details>` in Markdown, this may be helpful to some.

**Comment by waylan** [comment]
> I threw together a plugin called Spoilers

Interesting. Although it occurs to me that there is some existing syntax for "hiding spoilers." For example, the "movies" or "scifi" stack exchange sites use it as a way to hide "spoilers" for questions and answers regarding movies, etc. I mention it because I'm considered about creating multiple different syntax for the same end result. I realize your solution is specifically generating HTML that uses `<details><summary>`, which is not used by the existing implementations. But if the end goal is the same, do we need another alternative? Especially when the extension is called "Spoilers"?

**Comment by facelessuser** [comment]
> do we need another alternative?

I assume you are specifically referencing the syntax?  And if so, my personal answer is: I just let people decide for themselves if they find it useful.

There isn't currently an official spoiler syntax.  And there isn't really anything that currently does this in 3rd party Python Markdown extensions.  Also, I was unaware of any intention of one to be developed by you either. I guess you could say that stack exchange is common enough that maybe that syntax should be used.  I personally don't like the stack exchange syntax as it seems to piggy back off the blockquote syntax.  With things like SuperFences having to do coding gymnastics to get around blockquote syntax (since we don't currently have a good block handler for fences), I'm not too keen on adopting a variation of blockquotes that requires even more gymnastics for SuperFences.

There are other syntaxes available, but none ubiquitous enough for me to say that's common enough that people would instantly expect that.  Or maybe I'm just out of the loop.

FWIW it's just a name.  Details seemed boring and "Spoilers" resonated with me as you could easily use `details` for this purpose as well (some even essentially implement spoilers with `details` like behavior).  Eventually all browsers should support the `details` tag, so +1 for standards.  The current syntax is easy to work around for SuperFences and familiar to people who already use admonitions.

As with most of my extensions, I'm not trying to establish an official convention. Most of my extensions are born from my own selfish needs or preferences, but I don't mind sharing.  I don't really have a big audience to consult with about syntax and naming, or even a collaborative group.  I just kind of code up what moves me at the time for what I need in my documents and make it available; people use it or they don't.  Sure it's nice when others find it useful, and I do try to listen when I get feedback, but ultimately, I'm my end user.

With all of that said, I am open to suggestions to a point.  I don't think alternatives are inherently bad. It's a third party extension, so it's not an "official' anything.  Is it just the name that is found undesirable?  It's just been released, so maybe a course redirect could be made in regards to the name.  The name doesn't particular bother me, and it's not disingenuous  as "spoilers" is a reasonable usage for them.  If it's the idea of the extension as a whole...I'm not sure I'm looking to scrap it or greatly overhaul it, but I do try to keep open mind.  Do you have specific suggestions?

**Comment by facelessuser** [comment]
I will add, that if your future extension makes mine irrelevant, I have no problem deprecating mine.  I had been thinking about doing this extension for a while, but when there was some talk over on the mkdocs-material theme repo about this, I figured I'd finally offer a solution, especially since I have use for it.  I have no problem dropping extensions (I already have), I just  usually need to have my original requirements fulfilled in order for me to abandon what I have.  Less splintered  community means more focus on improving the one good extension etc.

**Comment by waylan** [comment]
@facelessuser You can create whatever extensions you want which implement whatever syntax you want. It doesn't matter to me. Personally, I have no need or interest in the feature. However, personally, I try to keep the introduction (by myself) of new syntax down to a minimum. If an existing syntax already exists that could serve my needs (no matter how uncommon it is), I personally would rather implement that in Python-Markdown than create something new. So I was just raising the question of whether another way to do the same thing is really necessary? To my eye, what you have looks more like an "accordion" than "spoiler" whereas there are existing "accordion" and "spoiler" implementations elsewhere and I expect people familiar with those might find your crossover surprising.

**Comment by facelessuser** [comment]
> To my eye, what you have looks more like an "accordion" than "spoiler" whereas there are existing "accordion" and "spoiler" implementations elsewhere and I expect people familiar with those might find your crossover surprising.

Maybe I was a little overly sensitive to the response. I better understand your statement now.

In retrospect, that is probably a better description opposed to spoilers even though some spoilers are accordions.  The name might be surprising to some, and probably not so surprising to others, but I do see better how this might be confusing.  Since this post, I've looked around at posts discussing syntaxes and see that there are people who are very passionate about conventions.  The spoiler discussion over at CommonMark is quite long.

Anyways, maybe a name change is appropriate moving forward since it is likely, that at some point, a common spoiler syntax may be present in the future.  Accordions may make more sense.  I did find some accordion syntaxes that messes with the header syntax (ugh), and some other elaborate variations that look to be more complicated than just using HTML, I will probably stick with the current syntax for now, but a name change seems to make sense as well.
